File: LeetCode/Problems/Python/74. Search a 2D Matrix.py
#Binary search
#Time: O(logn+logm)
#Space: O(1)
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or not matrix[0]:
            return False
        
        
        #binary search row
        l = 0
        r = len(matrix) -1
        row = -1
        while l <= r:
            mid = l + (r-l)//2
            if matrix[mid][0] <= target and matrix[mid][len(matrix[0])-1] >= target:
                row = mid
                break
            elif matrix[mid][0] < target:
                l = mid+1
            else:
                r = mid -1

        if row == -1:
            return False
        
        #binary search col
        l = 0 
        r = len(matrix[0]) -1
        
        while l <= r:
            mid = l + (r - l)//2
            if matrix[row][mid] == target:
                return True
            elif matrix[row][mid] < target:
                l = mid + 1
            elif matrix[row][mid] > target:
                r = mid - 1
        
        return False

File: LeetCode/Problems/C++/74. Search a 2D Matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(!matrix.size() || !matrix[0].size()) return false;
    
        //search for which row
        int lo = 0;
        int hi = matrix.size()-1;
        int mid;
        while(lo <= hi){
            mid = lo + (hi-lo)/2;
            if (matrix[mid][0] == target)
                return true;
            else if(matrix[mid][0] < target)
                lo = mid+1;
            else if(matrix[mid][0] > target)
                hi = mid-1;
        }
        
        int row = hi;
        
        if(row == -1 || row >= matrix.size()) return false;
        
        //search in the row
        lo = 0;
        hi = matrix[0].size()-1;
        
        while(lo <= hi){
            mid = lo + (hi-lo)/2;
            if (matrix[row][mid] == target)
                return true;
            else if(matrix[row][mid] < target)
                lo = mid+1;
            else
                hi = mid-1;
        }
        return false; 
    }
};